### ISSUE TITLE
Too many states are cached in memory when the state option EFFICIENT_STATE_MERGING is enabled

### ISSUE THREAD

**Issue by ltfish** [angr/angr#550]
With the new design `the_end_times` branch, we are essentially storing all states in the memory during symbolic execution. @rhelmot reported this issue. It should be investigated, and if it is the case, a new design is required to make `EFFICIENT_STATE_MERGING` to work.

**Comment by rhelmot** [comment]
dump from in-person conversation with yan: we don't really need the states for efficient state merging, we just need all the constraints that came from each path. Additionally, this option should be off by default, except for during veritesting.

**Comment by ltfish** [comment]
> we don't really need the states for efficient state merging, we just need all the constraints that came from each path.

I agree. However, spilling constraints out of states may not be optimal.

**Comment by zardus** [comment]
We already save off all added constraints in the StateHistory, so I don't think it'll really be any worse.
